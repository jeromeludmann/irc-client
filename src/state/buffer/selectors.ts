import { createSelector } from 'reselect'
import { selectBuffers } from '@app/state/server/selectors'
import { selectRoute } from '@app/state/route/selectors'

export const selectBuffer = createSelector(
  selectBuffers,
  selectRoute,
  (buffers, route) => buffers[route.bufferKey],
)

export const selectActivity = createSelector(
  selectBuffer,
  buffer => buffer.activity,
)
