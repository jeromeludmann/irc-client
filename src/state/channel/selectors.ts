import { createSelector } from "reselect";
import { selectChannel } from "@app/state/server/selectors";

export const selectMessages = createSelector(
  selectChannel,
  channel => channel.messages,
);

export const selectInput = createSelector(
  selectChannel,
  channel => channel.input,
);
