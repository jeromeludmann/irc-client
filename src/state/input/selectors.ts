import { createSelector } from 'reselect'
import { selectBuffer } from '@app/state/buffer/selectors'

export const selectInput = createSelector(
  selectBuffer,
  channel => channel.input,
)

export const selectInputValue = createSelector(
  selectInput,
  input => input.value,
)

export const selectInputDirtyValue = createSelector(
  selectInput,
  input => input.dirtyValue,
)

export const selectInputHistory = createSelector(
  selectInput,
  input => input.history,
)
