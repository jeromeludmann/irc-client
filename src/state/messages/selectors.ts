import { createSelector } from 'reselect'
import { selectBuffer } from '@app/state/buffer/selectors'

export const selectMessages = createSelector(
  selectBuffer,
  buffer => buffer.messages,
)
