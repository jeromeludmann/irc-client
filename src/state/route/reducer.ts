import { Action, AnyAction } from 'redux'
import { Route, BufferKey, isStatus } from '@app/utils/Route'
import {
  SWITCH_WINDOW,
  SwitchWindowAction,
  CLOSE_WINDOW,
} from '@app/actions/ui'
import {
  ReceiveJoinAction,
  RECEIVE_JOIN,
  ReceivePartAction,
  RECEIVE_PART,
} from '@app/actions/messages/incoming'
import { RootState } from '@app/state/root/reducer'

export type RouteState = Readonly<Route>

export const routeInitialState: RouteState = {
  serverKey: 'serverKey',
  bufferKey: BufferKey.STATUS,
}

type RouteReducer = (
  state: RouteState,
  action: Action,
  extraStates: { root: RootState },
) => RouteState

const itIsMe = (action: AnyAction, extraStates: { root: RootState }) =>
  action.payload.user.nick ===
  extraStates.root.servers[action.route.serverKey].user.nick

const caseReducers: { [action: string]: RouteReducer } = {
  [CLOSE_WINDOW]: (route, _, extraStates) => {
    if (isStatus(route.bufferKey)) {
      const keys = Object.keys(extraStates.root.servers)
      return keys.length > 1
        ? {
            serverKey: keys.filter(key => key !== route.serverKey)[0],
            bufferKey: BufferKey.STATUS,
          }
        : route
    }

    return { ...route, bufferKey: BufferKey.STATUS }
  },

  [RECEIVE_JOIN]: (route, action: ReceiveJoinAction, extraStates) =>
    itIsMe(action, extraStates) ? action.route : route,

  [RECEIVE_PART]: (route, action: ReceivePartAction, extraStates) =>
    itIsMe(action, extraStates)
      ? { ...route, bufferKey: BufferKey.STATUS }
      : route,

  [SWITCH_WINDOW]: (route, action: SwitchWindowAction, extraStates) => {
    const { serverKey, bufferKey } = action.route

    const routeFound =
      serverKey in extraStates.root.servers &&
      bufferKey in extraStates.root.servers[serverKey].buffers

    if (!routeFound) {
      console.log(`Route "${JSON.stringify(action.route)}" not found`)
      return route
    }

    return action.route
  },
}

export const reduceRoute: RouteReducer = (
  route = routeInitialState,
  action,
  extraStates,
) =>
  action.type in caseReducers
    ? caseReducers[action.type](route, action, extraStates)
    : route
