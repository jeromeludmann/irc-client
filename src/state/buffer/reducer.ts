import {
  MessagesState,
  messagesInitialState,
  reduceMessages,
} from '@app/state/messages/reducer'
import {
  InputState,
  inputInitialState,
  reduceInput,
} from '@app/state/input/reducer'
import { RoutedAction } from '@app/utils/Route'
import { RouteState } from '@app/state/route/reducer'
import { RECEIVE_JOIN, ReceiveJoinAction } from '@app/actions/messages/incoming'
import { SWITCH_WINDOW } from '@app/actions/ui'
import { ServerState } from '@app/state/server/reducer'
import { AnyAction } from 'redux'

type BufferPartialState = Readonly<{
  activity: boolean
}>

export type BufferState = Readonly<{
  input: InputState
  messages: MessagesState
}> &
  BufferPartialState

type BufferReducer<S = BufferState> = (
  buffer: S,
  action: RoutedAction,
  extraStates: {
    route: RouteState
    server: ServerState
  },
) => S

export const bufferInitialState = {
  activity: false,
  input: inputInitialState,
  messages: messagesInitialState,
}

const itIsMe = (action: AnyAction, extraStates: { server: ServerState }) =>
  action.payload.user.nick === extraStates.server.user.nick

const caseReducers: { [action: string]: BufferReducer<BufferPartialState> } = {
  [RECEIVE_JOIN]: (_, action: ReceiveJoinAction, extraStates) => ({
    activity: !itIsMe(action, extraStates),
  }),

  [SWITCH_WINDOW]: () => ({ activity: false }),
}

export const reduceBuffer: BufferReducer = (
  buffer,
  action,
  extraStates,
): BufferState => ({
  ...(action.type in caseReducers
    ? caseReducers[action.type](buffer, action, extraStates)
    : buffer),
  input: reduceInput(buffer.input, action),
  messages: reduceMessages(buffer.messages, action, extraStates),
})
