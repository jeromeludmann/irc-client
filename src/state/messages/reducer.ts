import { Action } from 'redux'
import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RawMessagesAction,
  CONNECTION_CLOSED,
  RAW_MESSAGES_RECEIVED,
  CONNECT_TO_SERVER,
  ConnectToServerAction,
} from '@app/actions/socket'
import {
  ReceiveErrorAction,
  ReceiveJoinAction,
  ReceivePartAction,
  ReceivePrivmsgAction,
  ReceiveNoticeFromServerAction,
  ReceiveNoticeFromChannelAction,
  ReceiveNoticeFromUserAction,
  RECEIVE_ERROR,
  RECEIVE_JOIN,
  RECEIVE_NOTICE_FROM_SERVER,
  RECEIVE_NOTICE_FROM_CHANNEL,
  RECEIVE_NOTICE_FROM_USER,
  RECEIVE_PART,
  RECEIVE_PRIVMSG,
  RECEIVE_PING_FROM_SERVER,
} from '@app/actions/messages/incoming'
import {
  SEND_PONG_TO_SERVER,
  SEND_PRIVMSG,
  SendPrivmsgAction,
} from '@app/actions/messages/outgoing'
import {
  PRINT_HELP_BY_DEFAULT,
  PrintHelpByDefaultAction,
  PRINT_HELP_ABOUT_COMMAND,
  PrintHelpAboutCommandAction,
} from '@app/actions/commands'
import { RouteState } from '@app/state/route/reducer'
import { ServerState } from '@app/state/server/reducer'
import { UPDATE_SERVER_LAG, UpdateServerLagAction } from '@app/actions/ui'

// TODO replace string[] by MessageState[]
export type MessagesState = string[]

type MessagesReducer = (
  messages: MessagesState,
  action: Action,
  extraStates: { route: RouteState; server: ServerState },
) => MessagesState

export const messagesInitialState: MessagesState = []

const caseReducers: { [action: string]: MessagesReducer } = {
  [CONNECT_TO_SERVER]: (messages, action: ConnectToServerAction) => [
    ...messages,
    `Connecting to ${action.payload.host}â€¦`,
  ],

  [CONNECTION_CLOSED]: messages => [
    ...messages,
    'Disconnected from remote host.',
  ],

  [CONNECTION_FAILED]: (
    messages,
    { payload: { message } }: ConnectionFailedAction,
  ) => [...messages, message],

  [PRINT_HELP_BY_DEFAULT]: (
    messages,
    { payload: { commands } }: PrintHelpByDefaultAction,
  ) => [
    ...messages,
    ...Object.keys(commands).map(
      commandName => `${commandName} - ${commands[commandName].description}`,
    ),
    'Type /HELP <command> for more details.',
  ],

  [PRINT_HELP_ABOUT_COMMAND]: (
    messages,
    { payload: { command } }: PrintHelpAboutCommandAction,
  ) => [
    ...messages,
    `Usage: /${command.name} ${command.syntax} - ${command.description}`,
  ],

  [RAW_MESSAGES_RECEIVED]: (messages, action: RawMessagesAction) => [
    ...messages,
    ...action.payload.messages,
  ],

  [RECEIVE_ERROR]: (messages, { payload: { message } }: ReceiveErrorAction) => [
    ...messages,
    message,
  ],

  [RECEIVE_JOIN]: (
    messages,
    { payload: { user, channel } }: ReceiveJoinAction,
  ) => [...messages, `${user.nick} has joined ${channel}`],

  [RECEIVE_NOTICE_FROM_SERVER]: (
    messages,
    action: ReceiveNoticeFromServerAction,
  ) => [...messages, action.payload.text],

  [RECEIVE_NOTICE_FROM_CHANNEL]: (
    messages,
    {
      payload: { user, text },
      route: { bufferKey },
    }: ReceiveNoticeFromChannelAction,
  ) => [...messages, `-${user.nick}/${bufferKey}- ${text}`],

  [RECEIVE_NOTICE_FROM_USER]: (
    messages,
    { payload: { user, text } }: ReceiveNoticeFromUserAction,
  ) => [...messages, `-${user.nick}- ${text}`],

  [RECEIVE_PART]: (
    messages,
    { payload: { user, channel, message } }: ReceivePartAction,
  ) => {
    const baseMsg = `${user.nick} has left ${channel}`
    return [...messages, message ? `${baseMsg} (${message})` : baseMsg]
  },

  [RECEIVE_PING_FROM_SERVER]: messages => [...messages, 'Ping?'],

  [RECEIVE_PRIVMSG]: (
    messages,
    { payload: { user, text } }: ReceivePrivmsgAction,
  ) => [...messages, `${user.nick}: ${text}`],

  [SEND_PONG_TO_SERVER]: messages => [...messages, 'Pong!'],

  [SEND_PRIVMSG]: (messages, action: SendPrivmsgAction, { server }) => [
    ...messages,
    `${server.user.nick}: ${action.payload.text}`,
  ],

  [UPDATE_SERVER_LAG]: (
    messages,
    { payload: { lag } }: UpdateServerLagAction,
    {},
  ) => [...messages, `Server lag: ${lag} ms`],
}

export const reduceMessages: MessagesReducer = (
  messagesState = messagesInitialState,
  action,
  extraStates,
) =>
  action.type in caseReducers
    ? caseReducers[action.type](messagesState, action, extraStates)
    : messagesState
