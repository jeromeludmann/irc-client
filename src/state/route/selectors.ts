import { RootState } from '@app/state/root/reducer'
import { RouteState } from '@app/state/route/reducer'

export const selectRoute = ({ route }: RootState): RouteState => route
