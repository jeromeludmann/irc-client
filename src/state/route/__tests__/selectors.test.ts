import { selectRoute } from '@app/state/route/selectors'
import { rootInitialState } from '@app/state/root/reducer'

describe('select route state', () => {
  it('should select route', () => {
    expect(selectRoute(rootInitialState)).toMatchSnapshot()
  })
})
