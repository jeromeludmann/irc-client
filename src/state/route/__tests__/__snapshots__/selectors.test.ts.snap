// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select route state should select route 1`] = `
Object {
  "bufferKey": "@status",
  "serverKey": "serverKey",
}
`;
